R/cube-transforms.R

```diff
@@ -0,0 +1,177 @@
+#' #                         18                         19                         35 
+#' #          Somewhat Disagree 
+#' #                         16 
+#' # to get the full array with the headings but not subtotals
+#' applyTransforms(crtabs(~opinion, ds), include = c("cube_cells", "headings")) 
+#' #               All opinions             Strongly Agree             Somewhat Agree 
+#' #                         NA                         23                         24 
+#' # Neither Agree nor Disagree          Strongly Disagree          Somewhat Disagree 
+#' #                         18                         19                         16 
+#' }
+#'
+#' @aliases subtotalArray
+#' @export
+applyTransforms <- function (x, ary = cubeToArray(x), ...) {
+    # if there are row transforms, calculate them and display them
+    row_trans <- tryCatch(Transforms(data=index(variables(x))[[1]]$view$transform), error = function(e) NULL)
+    if (!is.null(row_trans)) {
+        var_cats <- Categories(data=index(variables(x))[[1]]$categories)
+
+        # TODO: calculate category/element changes
+
+        if (length(dim(ary)) > 1) {
+            off_margins <- seq_along(dim(ary))[-1]
+            dim_names  <- names(dimnames(ary))
+            ary <- apply(ary, off_margins, calcTransforms, row_trans, var_cats, ...)
+            names(dimnames(ary)) <- dim_names
+        } else {
+            ary <- as.array(calcTransforms(ary, row_trans, var_cats, ...))
+        }
+
+        ary <- subsetTransformedCube(ary, x)
+    }
+
+    # TODO: calculate column transforms
+
+
+    return(ary)
+}
+
+subsetTransformedCube <- function (ary, cube) {
+    # subset variable categories to only include non-na
+    dims <- dim(ary)
+    keep_all <- lapply(seq_along(dims),
+                       function (i) {
+                           out <- rep(TRUE, dims[i])
+                           names(out) <- dimnames(ary)[[i]]
+                           return(out)
+                       })
+    names(keep_all) <- names(dimensions(cube))[seq_along(keep_all)]
+    keep_these_cube_dims <- evalUseNA(ary, dimensions(cube)[seq_along(keep_all)], cube@useNA)
+
+    # remove the categories determined to be removable above
+    keep_these <- mapply(function(x, y) {
+        not_ys <- y[!y]
+        x[names(not_ys)] <- not_ys
+        return(x)
+    }, keep_all,
+    keep_these_cube_dims,
+    SIMPLIFY=FALSE,
+    USE.NAMES=TRUE)
+
+    # match up those in keep_these that are already in ary
+    ary_dimnames <- dimnames(ary)
+    for (i in seq_along(ary_dimnames)) {
+        to_keep <- names(keep_these[[i]]) %in% ary_dimnames[[i]]
+        keep_these[[i]] <- keep_these[[i]][to_keep]
+    }
+
+    ary <- subsetCubeArray(ary, keep_these)
+
+    return(ary)
+}
+
+#' @rdname Transforms
+#' @export
+setMethod("transforms", "CrunchCube", function (x) {
+    trans <- tryCatch(Transforms(data=index(variables(x))[[1]]$view$transform), error = function(e) NULL)
```

Comment:
Hard-coding `[[1]]` means you're getting a `Transforms` object for the first dim. But let's think about where this is going. `lapply` over the variables (better: `dimensions`, see cube-variables.R for how `variables` isn't necessarily what you think it is), and return a list of Transforms. If there is none for that variable, what should it return? NULL or how about a 0-length Transforms object? If 0-length Transforms would just work (as in, your other code would handle that fine without the special-casing of is.null), maybe that's better.

Reply:
Anyway, make this return all the transforms for all dims, and then feel free to move your hard-coded [[1]] to the calling function(s) (applyTransforms)